R/standardize_MAE.R

```diff
@@ -298,21 +298,63 @@ set_unique_cl_names <- function(se) {
   checkmate::assert_class(se, "SummarizedExperiment")
   
   col_data <- SummarizedExperiment::colData(se)
+  col_data_new <- set_unique_cl_names_dt(col_data)
+  SummarizedExperiment::colData(se) <- col_data_new
+  
+  se
+}
+
+#' Set Unique Parental Identifiers in table
+#'
+#' This function sets the `CellLineName` field in 
+#' `colData` to be unique by appending the `clid` in parentheses for duplicates.
+#'
+#' @param col_data data.table or DFrame with row data
+#' @return fixed input table with unique `CellLineName` in `colData`.
+#' @examples
+#' col_data <- S4Vectors::DataFrame(CellLineName = c("ID1", "ID1"), clid = c("C1", "C2"))
+#' col_data <- set_unique_cl_names_dt(col_data)
+#' @export
+#' @keywords standardize_MAE
+#' 
+set_unique_cl_names_dt <- function(col_data, sep = " ") {
+  stopifnot(any(inherits(col_data, "data.table") || inherits(col_data, "DFrame")))
+  
   cellline_name <- get_env_identifiers("cellline_name")
   clid <- get_env_identifiers("cellline")
   
   if (!is.null(col_data[[cellline_name]])) {
-    duplicated_ids <- col_data[[cellline_name]][duplicated(col_data[[cellline_name]])]
+    if (data.table::is.data.table(col_data)) {
+      # for data.table check multiple columns; and update cl names only if adding clid suffix can make them different
+      unique_col_names <- c(unlist(get_default_identifiers()[
+        c("cellline_name", "drug_name", "drug_name2",
+          "concentration2", "duration", "data_source")
+      ]), "normalization_type")
+      unique_col_names <- intersect(unique_col_names, names(col_data))
+      duplicated_ids <- col_data[[cellline_name]][duplicated(col_data, by = unique_col_names)]
+      unique_col_names_clid <- c(unique_col_names, get_default_identifiers()$cellline)
+      duplicated_ids_with_clid <- col_data[[cellline_name]][duplicated(col_data, by = unique_col_names_clid)]
+      duplicated_ids <- setdiff(duplicated_ids, duplicated_ids_with_clid)
+    } else {
+      duplicated_ids <- col_data[[cellline_name]][duplicated(col_data[[cellline_name]])]
+    }
```

Comment:
Why there is a different procedure for colData for SE and for data.table?
(I understand that in colData we do not have information about Drug, but it is misleading: the same content as input but different output depending on format - in the context of presence of cell line name duplicates)

```
>   dt <- data.table::data.table(
   DrugName = c("DrugA", "DrugB", "DrugC", "DrugD", "DrugC", "DrugD"), 
   Gnumber = c("G1", "G2", "G3", "G4", "G3", "G4"),
   CellLineName = c("ID1", "ID1", "ID2", "ID2", "ID2", "ID2"), 
   clid = c("C1", "C2", "C3", "C4", "C5", "C6")
  )
>   res_dt <- set_unique_cl_names_dt(dt)
> res_dt
   DrugName Gnumber CellLineName   clid
     <char>  <char>       <char> <char>
1:    DrugA      G1          ID1     C1
2:    DrugB      G2          ID1     C2
3:    DrugC      G3     ID2 (C3)     C3
4:    DrugD      G4     ID2 (C4)     C4
5:    DrugC      G3     ID2 (C5)     C5
6:    DrugD      G4     ID2 (C6)     C6

  dt <- S4Vectors::DataFrame(
    DrugName = c("DrugA", "DrugB", "DrugC", "DrugD", "DrugC", "DrugD"), 
    Gnumber = c("G1", "G2", "G3", "G4", "G3", "G4"),
    CellLineName = c("ID1", "ID1", "ID2", "ID2", "ID2", "ID2"), 
    clid = c("C1", "C2", "C3", "C4", "C5", "C6")
  )
>   res_S4 <- set_unique_cl_names_dt(dt)
> res_S4
DataFrame with 6 rows and 4 columns
     DrugName     Gnumber CellLineName        clid
  <character> <character>  <character> <character>
1       DrugA          G1     ID1 (C1)          C1
2       DrugB          G2     ID1 (C2)          C2
3       DrugC          G3     ID2 (C3)          C3
4       DrugD          G4     ID2 (C4)          C4
5       DrugC          G3     ID2 (C5)          C5
6       DrugD          G4     ID2 (C6)          C6
```

Reply:
IMO now is complicated and not consistent.

The only thing to change is - instead of checking format of `col_data` input - just check whether `unique_col_names` is `"CellLineName"`.
If true - just add suffix, if not - check other columns and add suffix accordingly.
You have that code already written.

Reply:
I'm on the same page as @j-smola. It would be great to have consistent logic regardless of the data format.